### PR TITLE
Add Reverse Cuthill-McKee (RCM) placement algorithm

### PR DESCRIPTION
**Pull request by mossblaser**
This simple breadth-first-search-like sequential placer can be found in the software placement literature and is believed to perform a reasonable job of placing pipeline-like things.

Sorry to drop you in reviewing this clearly thesis-specific change @mundya ...

### CODE REVIEW

File: rig/place_and_route/place/rcm.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    for subgraph_vertices in _get_connected_subgraphs(vertices_resources,
+                                                      vertices_neighbours):
+        cm_order = _cuthill_mckee(subgraph_vertices, vertices_neighbours)
+        for vertex in reversed(cm_order):
+            yield vertex
+
+
+def rcm_chip_order(machine):
+    """A generator which iterates over a set of chips in a machine in
+    Reverse-Cuthil-McKee order.
```
Comment (mundya) [inline rig/place_and_route/place/rcm.py]: sp. Cuthill

File: rig/place_and_route/place/rcm.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    vertices_neighbours = _get_vertices_neighbours(nets)
+    for subgraph_vertices in _get_connected_subgraphs(vertices_resources,
+                                                      vertices_neighbours):
+        cm_order = _cuthill_mckee(subgraph_vertices, vertices_neighbours)
+        for vertex in reversed(cm_order):
+            yield vertex
+
+
+def rcm_chip_order(machine):
+    """A generator which iterates over a set of chips in a machine in
```
Comment (mundya) [inline rig/place_and_route/place/rcm.py]: The thing this returns yields vertices, not chips - or am I misreading?

File: tests/place_and_route/place/test_rcm.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+    # Seperate pairs of vertices
+    v0 = "v0"
+    v1 = "v1"
+    v2 = "v2"
+    v3 = "v3"
+    vertices_resources = {v0: {}, v1: {}, v2: {}, v3: {}}
+    nets = [Net(v0, v1), Net(v2, v3)]
+    assert list(rcm_vertex_order(vertices_resources, nets)) in (
+        # v1 & v2 first, all possible orders of each pair
```
Comment (mundya) [inline tests/place_and_route/place/test_rcm.py]: 0 & 1?

File: tests/place_and_route/place/test_rcm.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        [v0, v1, v2, v3],
+        [v0, v1, v3, v2],
+        [v1, v0, v2, v3],
+        [v1, v0, v3, v2],
+        # v3 & v4 first, all possible orders of each pair
```
Comment (mundya) [inline tests/place_and_route/place/test_rcm.py]: 2 & 3?

**Comment by mundya** [comment]
No test for the placer yet?

File: rig/place_and_route/place/rcm.py (hunk 2 above)
Comment (mossblaser) [inline rig/place_and_route/place/rcm.py]: You may be being misled: the implementation internally calls `rcm_vertex_order` to generate the ordering however it generates a graph in which the vertices are `(x, y)` chip coordinates and the nets represent working connections between those chips. As a result, the ordering produced by `rcm_vertex_order` is actually an ordering of chip coordinates.

Does that make sense? (What would be a better way of making this clearer in the code?)

File: tests/place_and_route/place/test_rcm.py (hunk 3 above)
Comment (mossblaser) [inline tests/place_and_route/place/test_rcm.py]: Thank you! Whoops!

**Comment by mundya** [comment]
Looks good to me (not sure I fully understand) apart from the comments above.

**Comment by mossblaser** [comment]
> No test for the placer yet?

There is a generic test suite for placement algorithms to which this placer has been added: https://github.com/project-rig/rig/pull/254/files#diff-e60413cd5d0cfb7adc8ce7a7f61c02e7R63 This suite should be sufficient, especially since this placer is really just a wrapper around the seqential placer in which the order in which vertices/chips are assigned is determined (in this placer) by the RCM algorthm.

File: rig/place_and_route/place/rcm.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    order.
+
+    For use as a vertex ordering for the sequential placer.
+    """
+    vertices_neighbours = _get_vertices_neighbours(nets)
+    for subgraph_vertices in _get_connected_subgraphs(vertices_resources,
+                                                      vertices_neighbours):
+        cm_order = _cuthill_mckee(subgraph_vertices, vertices_neighbours)
+        for vertex in reversed(cm_order):
+            yield vertex
```
Comment (mundya) [inline rig/place_and_route/place/rcm.py]: If `vertex` is a chip... then maybe it could be called `chip`?

Comment (mossblaser) [inline rig/place_and_route/place/rcm.py]: The `rcm_vertex_order` function's primary role is in fact for iterating over vertices in an application's graph. That said, as noted previously it is, in addition, internally used by `rcm_chip_order`.

Background: the `sequential` placer iterates over the vertices in an application graph and assigns them to chips in a machine, filling up one chip before moving on to the next. The order in which it iterates over vertices and fills chips is something the caller can control by providing an iterator over the vertices and chips respectively. The `breadth_first` placer simply provides a breadth-first-search-based iterator over vertices but doesn't specify a particular order in which to fill chips. The `hilbert` placer uses the same breadth-first vertex iterator used by the `breadth_first` placer but also provides an iterator over chips in a machine which causes the sequential placer to fill chips in hilbert-order.

The `rcm` placer simply defines that vertices should be placed in RCM order additionally that chips should be filled in RCM order. As such it implements two iterators `rcm_vertex_order` and `rcm_chip_order` which iterate over vertices in an application and chips in a machine (respectively) in RCM order. These iterators are then passed to the sequential placer which does the rest.

**Comment by mundya** [comment]
Cool, I think I understand this now, thank you!

**Comment by mundya** [comment]
Rigged

**Pull request merged**